Mathlib/Combinatorics/SimpleGraph/Diameter.lean

```diff
@@ -0,0 +1,208 @@
+/-
+Copyright (c) 2024 Rida Hamadani. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Rida Hamadani
+-/
+import Mathlib.Combinatorics.SimpleGraph.Metric
+
+/-!
+# Diameter of a simple graph
+
+This module defines the diameter of a simple graph, which measures the maximum distance between
+vertices.
+
+## Main definitions
+
+- `SimpleGraph.ediam` is the graph extended diameter.
+
+- `SimpleGraph.diam` is the graph diameter.
+
+## Todo
+
+- Prove that `G.egirth ≤ 2 * G.ediam + 1` and `G.girth ≤ 2 * G.diam + 1` when the diameter is
+  non-zero.
+
+-/
+
+namespace SimpleGraph
+variable {α : Type*} {G G' : SimpleGraph α}
+
+section ediam
+
+/--
+The extended diameter is the greatest distance between any two vertices, with the value `⊤` in
+case the distances are not bounded above, or the graph is not connected.
+-/
+noncomputable def ediam (G : SimpleGraph α) : ℕ∞ :=
+  ⨆ u, ⨆ v, G.edist u v
+
+lemma ediam_def : G.ediam = ⨆ p : (α × α), G.edist p.1 p.2 := by
+  rw [ediam, iSup_prod]
+
+lemma edist_le_ediam {u v : α} : G.edist u v ≤ G.ediam :=
+  (le_iSup _ _).trans <| le_iSup_iff.mpr fun _ h ↦ h u
```

Comment:
```suggestion
lemma edist_le_ediam {u v : α} : G.edist u v ≤ G.ediam := le_iSup₂ u v
```
In general, you should use `le_iSup_of_le` to chain, rather than `le_iSup`

Reply:
This is not working, maybe because the types of the parameters at `le_iSup₂` are different from what we have here? But I've used `le_iSup_of_le` for chaining, thanks!

Reply:
Ah sorry, `biSup_mono` is the lemma you want to be using here

Reply:
It is not clear to me how it applies here. Maybe I'm misunderstanding something, but `biSup_mono` seems to be a comparison between two `iSup`s, and the second parameter seems to be of type `prop` which depends on the first parameter.

Reply:
No, I was right. `le_iSup₂` is the correct lemma. Elaboration just seems to be weak
```suggestion
lemma edist_le_ediam {u v : α} : G.edist u v ≤ G.ediam := le_iSup₂ (f := G.edist) u v
```

Reply:
Ah, this notation is new to me, thank you very much!